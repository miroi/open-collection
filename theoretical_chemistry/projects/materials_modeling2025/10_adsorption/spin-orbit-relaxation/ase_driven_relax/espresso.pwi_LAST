&CONTROL
   verbosity        = 'low'
   tstress          = .true.
   tprnfor          = .true.
   outdir           = './tmp'
   prefix           = 'c_hg'
   pseudo_dir       = './'
/
&SYSTEM
   ibrav            = 0
   nat              = 9
   ntyp             = 2
   ecutwfc          = 65
   occupations      = 'smearing'
   degauss          = 0.01
   smearing         = 'gauss'
   noncolin         = .true.
   lspinorb         = .true.
   assume_isolated  = '2D'
   vdw_corr         = 'DFT-D3'
   starting_magnetization(1) = 0.0
   starting_magnetization(2) = 0.0
/
&ELECTRONS
   conv_thr         = 1e-10
/
&IONS
/
&CELL
/
&FCP
/
&RISM
/
ATOMIC_SPECIES
C  12.011 C_r.upf
Hg  200.592 Hg-sp_r.upf

K_POINTS automatic
2 2 1  0 0 0

CELL_PARAMETERS angstrom
4.91200017930000 0.00000000000000 0.00000000000000
-2.45600065610000 4.25391661160000 0.00000000000000
0.00000000000000 0.00000000000000 15.00000000000000

ATOMIC_POSITIONS angstrom
C  0.0000000000 0.0001481856 7.4565810343  
C  -1.2278718566 2.1268841769 7.4565810343  
C  2.4560000900 0.0000000000 7.4633449543  
C  1.2278712906 2.1268841769 7.4565810343  
C  1.2279309585 0.7090260158 7.4605167432  
C  -0.0000003780 2.8358648471 7.4605167432  
C  3.6840688855 0.7090260158 7.4605167432  
C  2.4559997120 2.8359442390 7.4512167563  
Hg  2.4559997120 2.8359442390 10.8341452861  

