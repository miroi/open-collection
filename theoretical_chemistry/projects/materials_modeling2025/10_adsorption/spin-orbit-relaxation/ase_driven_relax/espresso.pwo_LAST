
     Program PWSCF v.7.4.1 starts on 18Sep2025 at 22:31:35 
        Git branch: dftd4
        Last git commit: cdd2961ef42f8db591ced3ad6cdf9450b38a484e
        Last git commit date: Wed Jun 25 21:39:06 2025 +0300
        Last git commit subject: saving changes, mainly about ldftd4

     This program is part of the open-source Quantum ESPRESSO suite
     for quantum simulation of materials; please cite
         "P. Giannozzi et al., J. Phys.:Condens. Matter 21 395502 (2009);
         "P. Giannozzi et al., J. Phys.:Condens. Matter 29 465901 (2017);
         "P. Giannozzi et al., J. Chem. Phys. 152 154105 (2020);
          URL http://www.quantum-espresso.org", 
     in publications or presentations arising from this work. More details at
     http://www.quantum-espresso.org/quote

     Parallel version (MPI), running on     4 processors

     MPI processes distributed on     1 nodes
     6856 MiB available memory on the printing compute node when the environment starts

     Reading input from espresso.pwi
Warning: card &CELL ignored
Warning: card &FCP ignored
Warning: card &RISM ignored

     Current dimensions of program PWSCF are:
     Max number of different atomic species (ntypx) = 10
     Max number of k-points (npk) =  40000
     Max angular momentum in pseudopotentials (lmaxx) =  4
 set_vdw_corr : entering vdw_corr=DFT-D3                                                                          
     Message from routine set_vdw_corr:
     entering vdw_corr = DFT-D3
     Message from routine setup:
     using ibrav=0 with symmetry is DISCOURAGED, use correct ibrav instead

     R & G space division:  proc/nbgrp/npool/nimage =       4
     Subspace diagonalization in iterative solution of the eigenvalue problem:
     a serial algorithm will be used


     Parallelization info
     --------------------
     sticks:   dense  smooth     PW     G-vecs:    dense   smooth      PW
     Min         382     382    105                37390    37390    5379
     Max         383     383    106                37405    37405    5394
     Sum        1531    1531    421               149599   149599   21533

     Using Slab Decomposition



     bravais-lattice index     =            0
     lattice parameter (alat)  =       9.2823  a.u.
     unit-cell volume          =    2115.1209 (a.u.)^3
     number of atoms/cell      =            9
     number of atomic types    =            2
     number of electrons       =        52.00
     number of Kohn-Sham states=           62
     kinetic-energy cutoff     =      65.0000  Ry
     charge density cutoff     =     260.0000  Ry
     scf convergence threshold =      1.0E-10
     mixing beta               =       0.7000
     number of iterations used =            8  plain     mixing
     Exchange-correlation= PBE
                           (   1   4   3   4   0   0   0)
     Non magnetic calculation with spin-orbit


     celldm(1)=   9.282335  celldm(2)=   0.000000  celldm(3)=   0.000000
     celldm(4)=   0.000000  celldm(5)=   0.000000  celldm(6)=   0.000000

     crystal axes: (cart. coord. in units of alat)
               a(1) = (   1.000000   0.000000   0.000000 )  
               a(2) = (  -0.500000   0.866025   0.000000 )  
               a(3) = (   0.000000   0.000000   3.053746 )  

     reciprocal axes: (cart. coord. in units 2 pi/alat)
               b(1) = (  1.000000  0.577350 -0.000000 )  
               b(2) = (  0.000000  1.154701  0.000000 )  
               b(3) = (  0.000000 -0.000000  0.327467 )  


     PseudoPot. # 1 for C  read from file:
     ./C_r.upf
     MD5 check sum: d80ec9912146af6c4c1760317d1f95e9
     Pseudo is Norm-conserving + core correction, Zval =  4.0
     Generated using ONCVPSP code by D. R. Hamann
     Using radial grid of 1248 points,  6 beta functions with: 
                l(1) =   0
                l(2) =   0
                l(3) =   1
                l(4) =   1
                l(5) =   1
                l(6) =   1

     PseudoPot. # 2 for Hg read from file:
     ./Hg-sp_r.upf
     MD5 check sum: d85b3e284848a6ef6fce73cef7b59184
     Pseudo is Norm-conserving + core correction, Zval = 20.0
     Generated using ONCVPSP code by D. R. Hamann
     Using radial grid of 1246 points, 12 beta functions with: 
                l(1) =   0
                l(2) =   0
                l(3) =   1
                l(4) =   1
                l(5) =   1
                l(6) =   1
                l(7) =   2
                l(8) =   2
                l(9) =   2
               l(10) =   2
               l(11) =   3
               l(12) =   3

     atomic species   valence    mass     pseudopotential
     C                 4.00    12.01100     C ( 1.00)
     Hg               20.00   200.59200     Hg( 1.00)

      6 Sym. Ops. (no inversion) found ( 4 have fractional translation)



   Cartesian axes

     site n.     atom                  positions (alat units)
         1        C      tau(   1) = (   0.0000000   0.0000303   1.5180431  )
         2        C      tau(   2) = (  -0.2499738   0.4329975   1.5180431  )
         3        C      tau(   3) = (   0.5000000   0.0000000   1.5193415  )
         4        C      tau(   4) = (   0.2499736   0.4329975   1.5180431  )
         5        C      tau(   5) = (   0.2499858   0.1443458   1.5187978  )
         6        C      tau(   6) = (  -0.0000001   0.5773338   1.5187978  )
         7        C      tau(   7) = (   0.7500142   0.1443458   1.5187978  )
         8        C      tau(   8) = (   0.4999999   0.5773502   1.5170425  )
         9        Hg     tau(   9) = (   0.4999999   0.5773502   2.2056986  )

     --------------------------------------------
     DFT-D3 Dispersion Correction (3-body terms):
     --------------------------------------------
       Reference C6 values for interpolation: 

         atom   Coordination number   C6
         C          0.000            98.23
         C          0.987            86.49
         C          1.998            58.72
         C          2.999            51.56
         C          3.984            36.41
         Hg         0.000           724.15
         Hg         1.990           597.72

       Values used:

         atom   Coordination number  R0_AB[au]  C6      C8
         C          3.343            2.750     47.65   1378.25
         C          3.343            2.750     47.65   1378.25
         C          3.330            2.750     47.94   1386.64
         C          3.343            2.750     47.65   1378.25
         C          3.332            2.750     47.90   1385.46
         C          3.332            2.750     47.90   1385.46
         C          3.332            2.750     47.90   1385.46
         C          3.373            2.750     46.92   1357.02
         Hg         0.218            3.175    724.15  72691.35

         Molecular C6 ( Ry / a.u.^6 ) =      6712.34


     number of k points=     2  Gaussian smearing, width (Ry)=  0.0100
                       cart. coord. in units 2pi/alat
        k(    1) = (   0.0000000   0.0000000   0.0000000), wk =   0.2500000
        k(    2) = (   0.0000000  -0.5773503   0.0000000), wk =   0.7500000

     Dense  grid:   149599 G-vectors     FFT dimensions: (  48,  48, 150)

     Estimated max dynamical RAM per process >     154.83 MB

     Estimated total dynamical RAM >     619.32 MB
 ----2D----2D----2D----2D----2D----2D----2D----2D----2D----2D----2D----2D
  The code is running with the 2D cutoff
  Please refer to:
  Sohier, T., Calandra, M., & Mauri, F. (2017), 
  Density functional perturbation theory for gated two-dimensional heterostructures:
  Theoretical developments and application to flexural phonons in graphene.
  Physical Review B, 96(7), 75448. https://doi.org/10.1103/PhysRevB.96.075448
 ----2D----2D----2D----2D----2D----2D----2D----2D----2D----2D----2D----2D

     Initial potential from superposition of free atoms

     starting charge      51.9996, renormalised to      52.0000

     negative rho (up, down):  1.403E-05 0.000E+00
     Starting wfcs are   84 randomized atomic wfcs

     total cpu time spent up to now is        1.2 secs

     per-process dynamical memory:   118.9 Mb

     Self-consistent Calculation

     iteration #  1     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  1.00E-02,  avg # of iterations =  3.0

     total cpu time spent up to now is        3.9 secs

     total energy              =    -429.96176511 Ry
     estimated scf accuracy    <       0.77847244 Ry

     iteration #  2     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  1.50E-03,  avg # of iterations =  8.0

     total cpu time spent up to now is        6.7 secs

     total energy              =    -430.08262746 Ry
     estimated scf accuracy    <       0.05836622 Ry

     iteration #  3     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  1.12E-04,  avg # of iterations = 12.0

     total cpu time spent up to now is        9.7 secs

     total energy              =    -430.10103537 Ry
     estimated scf accuracy    <       0.06747506 Ry

     iteration #  4     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  1.12E-04,  avg # of iterations =  6.0

     total cpu time spent up to now is       11.9 secs

     total energy              =    -430.11909016 Ry
     estimated scf accuracy    <       0.00138527 Ry

     iteration #  5     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  2.66E-06,  avg # of iterations =  7.0

     total cpu time spent up to now is       14.9 secs

     total energy              =    -430.11945333 Ry
     estimated scf accuracy    <       0.00096148 Ry

     iteration #  6     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  1.85E-06,  avg # of iterations =  2.0

     total cpu time spent up to now is       17.3 secs

     total energy              =    -430.11963057 Ry
     estimated scf accuracy    <       0.00001530 Ry

     iteration #  7     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  2.94E-08,  avg # of iterations =  4.5

     total cpu time spent up to now is       20.0 secs

     total energy              =    -430.11963743 Ry
     estimated scf accuracy    <       0.00000368 Ry

     iteration #  8     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  7.08E-09,  avg # of iterations =  2.5

     total cpu time spent up to now is       22.2 secs

     total energy              =    -430.11963824 Ry
     estimated scf accuracy    <       0.00000041 Ry

     iteration #  9     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  7.83E-10,  avg # of iterations =  3.0

     total cpu time spent up to now is       24.4 secs

     total energy              =    -430.11963835 Ry
     estimated scf accuracy    <       0.00000002 Ry

     iteration # 10     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  4.74E-11,  avg # of iterations =  3.5

     total cpu time spent up to now is       26.9 secs

     total energy              =    -430.11963836 Ry
     estimated scf accuracy    <          5.6E-09 Ry

     iteration # 11     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  1.09E-11,  avg # of iterations =  2.0

     total cpu time spent up to now is       29.0 secs

     total energy              =    -430.11963836 Ry
     estimated scf accuracy    <          9.3E-10 Ry

     iteration # 12     ecut=    65.00 Ry     beta= 0.70
     Davidson diagonalization with overlap
     ethr =  1.78E-12,  avg # of iterations =  3.5

     total cpu time spent up to now is       31.0 secs

     End of self-consistent calculation

          k = 0.0000 0.0000 0.0000 ( 18643 PWs)   bands (ev):

  -117.6952-117.6952 -79.6879 -79.6879 -61.4214 -61.4214 -61.4203 -61.4203
   -23.9883 -23.9883 -18.6946 -18.6946 -18.6916 -18.6916 -18.6916 -18.6916
   -17.6927 -17.6927 -17.6920 -17.6920 -17.6920 -17.6920 -12.1070 -12.1070
   -10.8342 -10.8342 -10.8330 -10.8330 -10.8291 -10.8291 -10.8212 -10.8212
   -10.8114 -10.8114  -9.0035  -9.0035  -9.0018  -9.0018  -8.9808  -8.9808
    -7.3985  -7.3985  -7.3894  -7.3894  -7.1851  -7.1851  -6.7341  -6.7341
    -6.7337  -6.7337  -6.6673  -6.6673  -2.6834  -2.6834  -2.6715  -2.6715
    -2.6667  -2.6667  -1.3925  -1.3925   0.3200   0.3200

          k = 0.0000-0.5774 0.0000 ( 18732 PWs)   bands (ev):

  -117.6952-117.6952 -79.6879 -79.6879 -61.4214 -61.4214 -61.4203 -61.4203
   -22.5158 -22.5158 -22.5127 -22.5127 -19.6766 -19.6766 -19.6750 -19.6750
   -14.9342 -14.9342 -14.9337 -14.9337 -13.3916 -13.3916 -13.3908 -13.3908
   -11.8180 -11.8180 -11.8174 -11.8174 -10.8537 -10.8537 -10.8177 -10.8177
   -10.3725 -10.3725 -10.3023 -10.3023  -9.4381  -9.4381  -9.4379  -9.4379
    -9.0074  -9.0074  -8.9696  -8.9696  -8.9522  -8.9522  -7.1734  -7.1734
    -7.1181  -7.1181  -6.3744  -6.3744  -2.1114  -2.1114  -1.5030  -1.5030
    -1.4979  -1.4979  -0.2638  -0.2638   0.6808   0.6808

     the Fermi energy is    -5.5272 ev

!    total energy              =    -430.11963836 Ry
     estimated scf accuracy    <          2.2E-11 Ry
     smearing contrib. (-TS)   =      -0.00000000 Ry
     internal energy E=F+TS    =    -430.11963836 Ry

     The total energy is F=E-TS. E is the sum of the following terms:
     one-electron contribution =   -1842.88526123 Ry
     hartree contribution      =     916.90734908 Ry
     xc contribution           =     -89.63944045 Ry
     ewald contribution        =     585.54161701 Ry
     DFT-D3 Dispersion         =      -0.04390277 Ry

     convergence has been achieved in  12 iterations

     Forces acting on atoms (cartesian axes, Ry/au):

     atom    1 type  1   force =     0.00000000   -0.00000044   -0.00002332
     atom    2 type  1   force =    -0.00000038    0.00000022   -0.00002332
     atom    3 type  1   force =     0.00000000    0.00000000    0.00009315
     atom    4 type  1   force =     0.00000038    0.00000022   -0.00002332
     atom    5 type  1   force =     0.00000128   -0.00000074    0.00006784
     atom    6 type  1   force =    -0.00000000    0.00000148    0.00006784
     atom    7 type  1   force =    -0.00000128   -0.00000074    0.00006784
     atom    8 type  1   force =     0.00000000    0.00000000   -0.00022937
     atom    9 type  2   force =     0.00000000    0.00000000    0.00000264

     Total force =     0.000277     Total SCF correction =     0.000025


     Computing stress (Cartesian axis) and pressure

          total   stress  (Ry/bohr**3)                   (kbar)     P=        0.60
   0.00002154   0.00000000   0.00000000            3.17        0.00        0.00
   0.00000000   0.00002154   0.00000000            0.00        3.17        0.00
   0.00000000  -0.00000000  -0.00003077            0.00       -0.00       -4.53


     Writing all to output data dir ./tmp/c_hg.save/ :
     XML data file, charge density, pseudopotentials, collected wavefunctions

     init_run     :      1.04s CPU      1.16s WALL (       1 calls)
     electrons    :     29.06s CPU     29.82s WALL (       1 calls)
     forces       :      0.31s CPU      0.32s WALL (       1 calls)
     force_dftd3  :      0.07s CPU      0.07s WALL (       1 calls)
     stress       :      1.01s CPU      1.04s WALL (       1 calls)
     stres_dftd3  :      0.08s CPU      0.08s WALL (       1 calls)

     Called by init_run:
     wfcinit      :      0.93s CPU      1.02s WALL (       1 calls)
     potinit      :      0.05s CPU      0.06s WALL (       1 calls)
     hinit0       :      0.04s CPU      0.04s WALL (       1 calls)

     Called by electrons:
     c_bands      :     24.41s CPU     24.95s WALL (      12 calls)
     sum_band     :      3.57s CPU      3.71s WALL (      12 calls)
     v_of_rho     :      0.44s CPU      0.47s WALL (      13 calls)
     mix_rho      :      0.54s CPU      0.59s WALL (      12 calls)
     energy_dftd3 :      0.06s CPU      0.06s WALL (       1 calls)

     Called by c_bands:
     init_us_2    :      0.28s CPU      0.29s WALL (      54 calls)
     cegterg      :     22.76s CPU     23.30s WALL (      24 calls)

     Called by *egterg:
     cdiaghg      :      0.34s CPU      0.35s WALL (     138 calls)
     h_psi        :     19.02s CPU     19.59s WALL (     140 calls)
     g_psi        :      0.08s CPU      0.08s WALL (     114 calls)

     Called by h_psi:
     h_psi:calbec :      1.33s CPU      1.34s WALL (     140 calls)
     vloc_psi     :     16.29s CPU     16.82s WALL (     140 calls)
                                        0.00s GPU  (     140 calls)
     add_vuspsi   :      1.28s CPU      1.29s WALL (     140 calls)

     General routines
     calbec       :      2.04s CPU      2.06s WALL (     174 calls)
     fft          :      0.47s CPU      0.51s WALL (     229 calls)
     ffts         :      0.07s CPU      0.07s WALL (      48 calls)
     fftw         :     16.23s CPU     16.77s WALL (   20176 calls)

     Parallel routines

     PWSCF        :     31.49s CPU     32.47s WALL


   This run was terminated on:  22:32: 8  18Sep2025            

=------------------------------------------------------------------------------=
   JOB DONE.
=------------------------------------------------------------------------------=
