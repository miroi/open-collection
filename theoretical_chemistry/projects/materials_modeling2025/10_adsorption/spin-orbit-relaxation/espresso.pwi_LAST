&CONTROL
   verbosity        = 'low'
   tstress          = .true.
   tprnfor          = .true.
   outdir           = './tmp'
   prefix           = 'c_hg'
   pseudo_dir       = './'
/
&SYSTEM
   ibrav            = 0
   nat              = 9
   ntyp             = 2
   ecutwfc          = 65
   occupations      = 'smearing'
   degauss          = 0.01
   smearing         = 'gauss'
   noncolin         = .true.
   lspinorb         = .true.
   assume_isolated  = '2D'
   vdw_corr         = 'DFT-D3'
   starting_magnetization(1) = 0.0
   starting_magnetization(2) = 0.0
/
&ELECTRONS
   conv_thr         = 1e-10
/
&IONS
/
&CELL
/
&FCP
/
&RISM
/
ATOMIC_SPECIES
C  12.011 C_r.upf
Hg  200.592 Hg-sp_r.upf

K_POINTS automatic
2 2 1  0 0 0

CELL_PARAMETERS angstrom
4.91200017930000 0.00000000000000 0.00000000000000
-2.45600065610000 4.25391661160000 0.00000000000000
0.00000000000000 0.00000000000000 15.00000000000000

ATOMIC_POSITIONS angstrom
C  0.0000000000 0.0001490009 7.4566279210  
C  -1.2278711386 2.1268837714 7.4566279210  
C  2.4560000900 0.0000000000 7.4630059115  
C  1.2278705726 2.1268837714 7.4566279210  
C  1.2279301724 0.7090264656 7.4603351987  
C  -0.0000003780 2.8358639294 7.4603351987  
C  3.6840696716 0.7090264656 7.4603351987  
C  2.4559997120 2.8359442390 7.4517131370  
Hg  2.4559997120 2.8359442390 10.8343919637  

