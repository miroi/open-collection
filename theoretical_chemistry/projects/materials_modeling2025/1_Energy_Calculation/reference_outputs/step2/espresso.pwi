&CONTROL
   calculation      = 'scf'
   pseudo_dir       = './'
/
&SYSTEM
   ibrav            = 0
   nat              = 2
   ntyp             = 1
   ecutwfc          = 30
   occupations      = 'smearing'
   degauss          = 0.01
   smearing         = 'gauss'
/
&ELECTRONS
   conv_thr         = 1e-10
/
&IONS
/
&CELL
/
&FCP
/
&RISM
/
ATOMIC_SPECIES
Si 28.085 Si.upf

K_POINTS automatic
2 2 2  0 0 0

CELL_PARAMETERS angstrom
0.00000000000000 2.71490000000000 2.71490000000000
2.71490000000000 0.00000000000000 2.71490000000000
2.71490000000000 2.71490000000000 0.00000000000000

ATOMIC_POSITIONS angstrom
Si 1.3574500000 4.0723500000 4.0723500000  
Si 0.0000000000 0.0000000000 0.0000000000  

