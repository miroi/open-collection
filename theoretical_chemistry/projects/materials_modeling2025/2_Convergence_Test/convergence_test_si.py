#!/usr/bin/env python
from ase import Atoms
from ase.calculators.espresso import Espresso, EspressoProfile
import os
import sys
import numpy as np
sys.stdout.flush()

# ==============================================
# ASE Convergence test (Cutoff and K-points)
# ==============================================

# ==============================================
# 1. Quantum Espresso input parameters
# ==============================================
pseudopotentials = {'Si': 'Si.upf'} 

# Base SCF input parameters
base_input_data = {
    'control': {
        'calculation': 'scf',
        'prefix': 'si',
        'outdir': './tmp',
        'verbosity': 'low'
    },
    'system': {
        'occupations': 'smearing',
        'smearing': 'gauss',
        'degauss': 0.01,
        'ibrav': 0,
        'nat': 2,
        'ntyp': 1
    },
    'electrons': {
        'conv_thr': 1.0e-10
    }
}

# ==============================================
# 2. Atomic Structure 
# ==============================================
atoms = Atoms(
    symbols=['Si']*2,
    positions=[
        [1.3574500000, 4.0723500000, 4.0723500000],
        [0.0000000000, 0.0000000000, 0.0000000000]
    ],
    cell=[
        [0.0000000000, 2.7149000000, 2.7149000000],
        [2.7149000000, 0.0000000000, 2.7149000000],
        [2.7149000000, 2.7149000000, 0.0000000000]
    ],
    pbc=[True, True, True]
)

# ==============================================
# 3. Calculator configuration
# ==============================================
# Set QE bin directory 
#qe_bin = "/home/dsen/work/bin/qe-7.4.1_serial"
qe_bin = "/home/dsen/work/bin/qe-7.4.1"

# Job commands
#pw_command = f'{qe_bin}/bin/pw.x'
pw_command = f'mpirun -np 4 {qe_bin}/bin/pw.x'

pw_profile = EspressoProfile(
    command=pw_command,
    pseudo_dir='./'
)

# Test values
ecutwfc_values = np.arange(20, 120, 5)  
kpoints_values = [(k,k,k) for k in range(2, 25)] 

# Convergence threshold (meV/atom)
energy_tol = 0.01  

# ==============================================
# 4. Run Convergence Tests
# ==============================================
def run_convergence_test():
    results = []
    natoms = base_input_data['system']['nat']  
    
    # Determine intermediate k-point grid for cutoff testing (ensuring integer values)
    mid_index = len(kpoints_values) // 2
    k_intermediate = kpoints_values[mid_index]
    print(f"\nUsing k-point grid {k_intermediate} for ecutwfc convergence testing", flush=True)
    
    print("\nStarting convergence tests...", flush=True)
    print("="*60, flush=True)
    
    # 1. ecutwfc convergence testing
    print("\n1. Testing ecutwfc convergence with k-point grid", k_intermediate, ":", flush=True)
    print("-"*60, flush=True)
    print("ecutwfc (Ry)   Energy (eV)    ΔE (meV/atom)", flush=True)
    print("-"*60, flush=True)
    
    prev_energy = None
    converged_ecut = None
    
    for ecut in ecutwfc_values:
        input_data = base_input_data.copy()
        input_data['system']['ecutwfc'] = int(ecut)
        
        calc = Espresso(
            profile=pw_profile,
            pseudopotentials=pseudopotentials,
            input_data=input_data,
            kpts=k_intermediate)
        atoms.calc = calc
        energy = atoms.get_potential_energy()
        
        if prev_energy is not None:
            delta_e = (energy - prev_energy)/natoms * 1000  # meV/atom
            converged = "*" if abs(delta_e) < energy_tol else ""
            print(f"{ecut:10.1f}    {energy:12.6f}   {delta_e:8.3f} {converged}", flush=True)
            
            if abs(delta_e) < energy_tol and converged_ecut is None:
                converged_ecut = ecut
                print("-"*60, flush=True)
                print(f">>> ecutwfc CONVERGED at {ecut} Ry (ΔE < {energy_tol} meV/atom)", flush=True)
                print("-"*60, flush=True)
                break
        else:
            print(f"{ecut:10.1f}    {energy:12.6f}     -", flush=True)
        
        results.append(('ecut', ecut, k_intermediate, energy, delta_e if prev_energy is not None else 0))
        prev_energy = energy
    
    # 2. k-point convergence testing
    print("\n2. Testing k-point convergence with ecutwfc =", 
          f"{converged_ecut if converged_ecut else ecutwfc_values[-1]} Ry" +
          f"{' (using max value)' if not converged_ecut else ''}:", flush=True)
    print("-"*60, flush=True)
    print("K-point grid   Energy (eV)    ΔE (meV/atom)", flush=True)
    print("-"*60, flush=True)
    
    prev_k_energy = None
    converged_kpts = None
    
    for kpts in kpoints_values:
        input_data = base_input_data.copy()
        input_data['system']['ecutwfc'] = int(converged_ecut if converged_ecut else ecutwfc_values[-1])
        
        calc = Espresso(
            profile=pw_profile,
            pseudopotentials=pseudopotentials,
            input_data=input_data,
            kpts=kpts)
        atoms.calc = calc
        energy = atoms.get_potential_energy()
        
        if prev_k_energy is not None:
            delta_e = (energy - prev_k_energy)/natoms * 1000  # meV/atom
            converged = "*" if abs(delta_e) < energy_tol else ""
            print(f"{str(kpts):12s}    {energy:12.6f}   {delta_e:8.3f} {converged}", flush=True)
            
            if abs(delta_e) < energy_tol and converged_kpts is None:
                converged_kpts = kpts
                print("-"*60, flush=True)
                print(f">>> K-points CONVERGED at {kpts} (ΔE < {energy_tol} meV/atom)", flush=True)
                print("-"*60, flush=True)
                break
        else:
            print(f"{str(kpts):12s}    {energy:12.6f}     -", flush=True)
        
        results.append(('kpts', converged_ecut if converged_ecut else ecutwfc_values[-1], 
                       kpts, energy, delta_e if prev_k_energy is not None else 0))
        prev_k_energy = energy
    
    print("="*60, flush=True)
    print("Convergence summary:", flush=True)
    print(f"- ecutwfc: {converged_ecut if converged_ecut else 'Not converged (using max value)'}", flush=True)
    print(f"- k-points: {converged_kpts if converged_kpts else 'Not converged'}", flush=True)
    print("="*60, flush=True)
    
    return results

# Run the tests
convergence_results = run_convergence_test()