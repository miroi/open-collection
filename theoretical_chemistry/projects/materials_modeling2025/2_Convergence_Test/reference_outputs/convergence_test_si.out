
Using k-point grid (13, 13, 13) for ecutwfc convergence testing

Starting convergence tests...
============================================================

1. Testing ecutwfc convergence with k-point grid (13, 13, 13) :
------------------------------------------------------------
ecutwfc (Ry)   Energy (eV)    ΔE (meV/atom)
------------------------------------------------------------
      20.0     -230.247524     -
      25.0     -230.266372     -9.424 
      30.0     -230.271329     -2.479 
      35.0     -230.273691     -1.181 
      40.0     -230.274387     -0.348 
      45.0     -230.274473     -0.043 
      50.0     -230.274554     -0.041 
      55.0     -230.274646     -0.046 
      60.0     -230.274690     -0.022 
      65.0     -230.274700     -0.005 *
------------------------------------------------------------
>>> ecutwfc CONVERGED at 65 Ry (ΔE < 0.01 meV/atom)
------------------------------------------------------------

2. Testing k-point convergence with ecutwfc = 65 Ry:
------------------------------------------------------------
K-point grid   Energy (eV)    ΔE (meV/atom)
------------------------------------------------------------
(2, 2, 2)        -227.715356     -
(3, 3, 3)        -229.655202   -969.923 
(4, 4, 4)        -230.086156   -215.477 
(5, 5, 5)        -230.210060    -61.952 
(6, 6, 6)        -230.250699    -20.320 
(7, 7, 7)        -230.265272     -7.287 
(8, 8, 8)        -230.270850     -2.789 
(9, 9, 9)        -230.273095     -1.123 
(10, 10, 10)     -230.274034     -0.470 
(11, 11, 11)     -230.274439     -0.203 
(12, 12, 12)     -230.274619     -0.090 
(13, 13, 13)     -230.274700     -0.041 
(14, 14, 14)     -230.274738     -0.019 
(15, 15, 15)     -230.274756     -0.009 *
------------------------------------------------------------
>>> K-points CONVERGED at (15, 15, 15) (ΔE < 0.01 meV/atom)
------------------------------------------------------------
============================================================
Convergence summary:
- ecutwfc: 65
- k-points: (15, 15, 15)
============================================================
